roles/openshift_node/files/bootstrap.yml

```diff
@@ -0,0 +1,69 @@
+#!/usr/bin/ansible-playbook
+---
+- hosts: localhost
+  gather_facts: yes
+  vars:
+    origin_dns:
+      file: /etc/dnsmasq.d/origin-dns.conf
+      lines:
+      - regex: ^listen-address
+        state: present
+        line: "listen-address={{ ansible_default_ipv4.address }}"
+    node_dns:
+      file: /etc/dnsmasq.d/node-dnsmasq.conf
+      lines:
+      - regex: "^server=/in-addr.arpa/127.0.0.1$"
+        line: server=/in-addr.arpa/127.0.0.1
+      - regex: "^server=/cluster.local/127.0.0.1$"
+        line: server=/cluster.local/127.0.0.1
+
+  tasks:
+  - include_vars: openshift_settings.yaml
+
+  - name: set the data for node_dns
+    lineinfile:
+      create: yes
+      insertafter: EOF
+      path: "{{ node_dns.file }}"
+      regexp: "{{ item.regex }}"
+      line: "{{ item.line | default(omit) }}"
+    with_items: "{{ node_dns.lines }}"
+
+  - name: set the data for origin_dns
+    lineinfile:
+      create: yes
+      state: "{{ item.state | default('present') }}"
+      insertafter: "{{ item.after | default(omit) }}"
+      path: "{{ origin_dns.file }}"
+      regexp: "{{ item.regex }}"
+      line: "{{ item.line | default(omit)}}"
+    with_items: "{{ origin_dns.lines }}"
+
+  - when:
+    - openshift_group_type is defined
+    - openshift_group_type != ''
+    - openshift_group_type != 'master'
+    block:
+    - name: determine the openshift_service_type
+      stat:
+        path: /etc/sysconfig/atomic-openshift-node
+      register: service_type_results
+
+    - name: set openshift_service_type fact based on stat results
+      set_fact:
+        openshift_service_type: "{{ service_type_results.stat.exists | ternary('atomic-openshift', 'origin') }}"
+
+    - name: update the sysconfig to have necessary variables
+      lineinfile:
+        dest: "/etc/sysconfig/{{ openshift_service_type }}-node"
+        line: "{{ item.line }}"
+        regexp: "{{ item.regexp }}"
+      with_items:
+      - line: "BOOTSTRAP_CONFIG_NAME=node-config-{{ openshift_group_type }}"
+        regexp: "^BOOTSTRAP_CONFIG_NAME=.*"
+
+    - name: update the ExecStart to have bootstrap
```

Comment:
We could probably do this in the AMI build.

Reply:
If we generate a specific node AMI (master|compute|infra) then I agree.  We aren't currently planning on doing this unless something has changed.

Until we build separate AMIs for each node type then I'm not sure how we do this at AMI build time.  If I missed something please explain.

Reply:
I was under the assumption that bootstrapping was generic enough that each node could get it's config from the master, determined based on what the master needs.  I don't see a reason to have to assign node types in this way if we're going to get our config from the master.

Also, what about just forcing the nodes bootstrapped via AMI to be compute types?  We can dictate that masters are infra.

Reply:
> I was under the assumption that bootstrapping was generic enough that each node could get it's config from the master, determined based on what the master needs. I don't see a reason to have to assign node types in this way if we're going to get our config from the master.

The master doesn't know what group type is desired when starting up unless you tell it.  This could be done through instance tags which would be great but isn't implemented that way.  As the code stands today, we need to tell the node what type they are.  Is there another way of doing this?

> Also, what about just forcing the nodes bootstrapped via AMI to be compute types? We can dictate that masters are infra.

This would simplify our process but we do not run any pods on the masters today.  This is done for a number of reasons.  I'll name a few:
- resource consumption of master-api, master-controller, and etcd running on masters
- security.  In the last 2 weeks we've had two separate CVE's that have had potential to escalate privileges out of containers.  If master were compromised then we are in big trouble.
- port collisions with master-api and the router

Reply:
My thinking is that when the nodes contact the master for their configs, the master says 'oh, I need another infra node, next node to enroll is infra node.'

Ideally, when the node joins the cluster, it can say "I have x features, ram, etc, my tags/what I found in my local config is y" and master can act on that data to supply the correct parameters to the node.

> This would simplify our process but we do not run any pods on the masters today. This is done for a number of reasons. I'll name a few:

Those are really good points.  We should get some of this captured into our install docs, as it doesn't come with this type of warning against running infra pods on masters.

Reply:
That may be something in the future, but for now the group determines the configuration.  Groups have security rules, firewalls, load balancers, and other characteristics that are determined outside the node.  Static configuration defining those on group is the direction we're going for this and for Kube.

The master can't make the determination what security groups the nodes are in, because the master won't necessarily have access to that info.